intel-sgx/ppid-decryption-tool/Enclave/id_enclave.c

```diff
@@ -0,0 +1,252 @@
+                                         (unsigned char*)g_rsa_key.n,
+                                         (unsigned char*)g_rsa_key.d,
+                                         (unsigned char*)g_rsa_key.e,
+                                         (unsigned char*)g_rsa_key.p,
+                                         (unsigned char*)g_rsa_key.q,
+                                         (unsigned char*)g_rsa_key.dmp1,
+                                         (unsigned char*)g_rsa_key.dmq1,
+                                         (unsigned char*)g_rsa_key.iqmp);
+    if (sgx_status != SGX_SUCCESS) {
+        return sgx_status;
+    }
+
+    // PCE wants the key in big endian
+    size_t i;
+    uint8_t* p_temp;
+    p_temp = (uint8_t*)g_rsa_key.e;
+    for (i = 0; i < REF_RSA_OAEP_3072_EXP_SIZE; i++) {
+        p_rsa_pub_key->e[i] = *(p_temp + REF_RSA_OAEP_3072_EXP_SIZE - 1 - i); //create big endian e
+    }
+    p_temp = (uint8_t*)g_rsa_key.n;
+    for (i = 0; i < REF_RSA_OAEP_3072_MOD_SIZE; i++) {
+        p_rsa_pub_key->n[i] = *(p_temp + REF_RSA_OAEP_3072_MOD_SIZE - 1 - i); //create big endian n
+    }
+
+    do {
+        sgx_status = sgx_sha256_init(&sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+
+        sgx_status = sgx_sha256_update(&crypto_suite,
+            sizeof(uint8_t),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        //(MOD followed by e)
+        sgx_status = sgx_sha256_update(p_rsa_pub_key->n,
+            sizeof(p_rsa_pub_key->n),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        sgx_status = sgx_sha256_update(p_rsa_pub_key->e,
+            sizeof(p_rsa_pub_key->e),
+            sha_handle);
+        if (SGX_SUCCESS != sgx_status)
+            break;
+        // !!! This is the translation of C++ code into C code. The previous code looked like this:
+        // sgx_status = sgx_sha256_get_hash(sha_handle, (sgx_sha256_hash_t *)reinterpret_cast<sgx_sha256_hash_t *>(&report_data));
+        sgx_status = sgx_sha256_get_hash(sha_handle, (sgx_sha256_hash_t *)&report_data);
+    } while (0);
+    if (SGX_SUCCESS != sgx_status) {
+        if (SGX_ERROR_OUT_OF_MEMORY != sgx_status)
+            sgx_status = SGX_ERROR_UNEXPECTED;
+        goto ret_point;
+    }
+
+    sgx_status = sgx_create_report(p_pce_target_info, &report_data, p_ide_report);
+    if (SGX_SUCCESS != sgx_status && SGX_ERROR_OUT_OF_MEMORY != sgx_status) {
+        sgx_status = SGX_ERROR_UNEXPECTED;
+    }
+
+    ret_point:
+    // Clear critical output data on error
+    if (SGX_SUCCESS != sgx_status) {
+        memset_s(p_ide_report, sizeof(*p_ide_report), 0, sizeof(*p_ide_report));
+    }
+    if (sha_handle != NULL) {
+        sgx_sha256_close(sha_handle);
+    }
+
+    return sgx_status;
+}
+
+sgx_status_t ide_decrypt_ppid(uint32_t encrypted_ppid_size, uint8_t *p_encrypted_ppid, uint8_t* ppid)
```

Comment:
Add function comment headers like its done for ide_get_pce_encrypt_key. Also add comments about each parameter - for example ppid is expected to be initialized to zero and have a minimum size of DECRYPTED_PPID_LENGTH.

It is also beneficial to do validation checks on the function parameters at the start of every function - for example ensure pointers aren't null.